### PR TITLE
cluster up: do not re-initialize a cluster that already has been initialized

### PR DESCRIPTION
**Pull request by csrwng**
A condition function can now be attached to a bootstrap task so that if the condition evaluates to false, the task is skipped altogether.

### CODE REVIEW

**Comment by csrwng** [comment]
@bparees ptal

File: pkg/bootstrap/docker/up.go [hunk 1]
```diff
@@ -873,3 +887,67 @@ func (c *ClientStartConfig) determineIP(out io.Writer) (string, error) {
...
+	}
+
+	result := func() bool {
+		if !c.UseExistingConfig {
+			return true
+		}
+
+		cfg, _, err := c.OpenShiftHelper().GetConfig(c.LocalConfigDir)
+		if err != nil {
+			glog.V(2).Infof("error reeading config: %v", err)
```
Comment (bparees) [inline pkg/bootstrap/docker/up.go]: reading.

**Comment by csrwng** [comment]
@bparees the use case is that across different runs of cluster up you may be preserving your data and config, and may not want cluster up to undo changes you made. For example, you could have deleted the default project, replaced default imagestreams, or changed the user provider.

**Comment by jorgemoralespou** [comment]
@bparees this is needed

**Comment by bparees** [comment]
> @bparees this is needed

only if we want to encourage using oc cluster up for maintaining "long term" environments.  I think we should at least stop and discuss that before plowing ahead.  I'm not saying it's not valid, but i wonder what sort of a slippery slope we're on before people are asking for oc cluster up to launch multiple nodes, or run multimaster, or otherwise supplant ansible for managing a cluster?

@smarterclayton wdyt?

**Comment by smarterclayton** [comment]
If the origin container exists we should stop immediately.  Is this trying to reapply config?

**Comment by csrwng** [comment]
@smarterclayton this is for the use case when you are reusing config or the data directory (starting cluster up again). If you've already initialized your data, there's no need to re-initialize it.

File: pkg/bootstrap/docker/up.go (hunk 1 above)
Comment (csrwng) [inline pkg/bootstrap/docker/up.go]: fixed

**Comment by csrwng** [comment]
@bparees I wouldn't necessarily call it a long-term use case to change the default config then start cluster up again to try out your config change. I would think that's something we need for the developer use case. Also, this change is not adding any function to cluster up, simply preventing the blind initialization of data when you already have an existing data/config you want to use.

**Comment by smarterclayton** [comment]
If you have a container already, you should use docker start origin.

On Sep 28, 2016, at 9:07 PM, Cesar Wong <notifications@github.com> wrote:

@bparees <https://github.com/bparees> I wouldn't necessarily call it a
long-term use case to change the default config then start cluster up again
to try out your config change. I would think that's something we need for
the developer use case. Also, this change is not adding any function to
cluster up, simply preventing the blind initialization of data when you
already have an existing data/config you want to use.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/openshift/origin/pull/11146#issuecomment-250344771>,
or mute
the thread

**Comment by csrwng** [comment]
> If you have a container already, you should use docker start origin.

You don't have a container. All you have is config and data.

**Comment by smarterclayton** [comment]
I am extremely skeptical of this use case.  Let's talk tomorrow

On Sep 28, 2016, at 9:23 PM, Cesar Wong <notifications@github.com> wrote:

If you have a container already, you should use docker start origin.

You don't have a container. All you have is config and data.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/openshift/origin/pull/11146#issuecomment-250346776>,
or mute
the thread

**Comment by jorgemoralespou** [comment]
@smarterclayon @bparees @cswong this does not add any additional use case. Currently there's a switch --keep-config. This PR just makes that switch consistent. If you keep configuration between restarts of the clusters shouldn't be initializing again stuff, as a) it shouldn't be needed and b) if a user has deleted changed some config why bringing back up the cluster should change user decisions?

**Comment by csrwng** [comment]
@jorgemoralespou yes, I guess I didn't explain that well enough... will talk with @smarterclayton later today. And here I thought this was the least controversial out of any of the 'cluster up' fixes :-)

**Comment by jorgemoralespou** [comment]
@cswong, @smarterclayton should change his nick to fightingclayton or opinnionatedclayton. :-D

**Comment by bparees** [comment]
[merge]

**Comment by csrwng** [comment]
Flake https://github.com/openshift/origin/issues/11094
[test]

**Pull request merged**